text/0000-bom-in-layer-metadata.md

```diff
@@ -0,0 +1,106 @@
+# Meta
+[meta]: #meta
+- Name: BOM inclusion in layer content metadata
+- Start Date: 2021-04-01
+- Author(s): [@samj1912](https://github.com/samj1912)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes the inclusion of a `BOM` (Bill-of-materials) table in layer content metadata TOML files (`<layer-name.toml>`). This would allow the `BOM`s generated from a layer to have the same lifetime as the layer itself and avoid various gotchas that happen currently as a result of the BOM being defined in the `launch.toml` or `build.toml`.
+
+# Motivation
+[motivation]: #motivation
+
+## Why should we do this?
+
+Currently `BOM` is a buildpack level entity and has to be defined in either the `launch.toml` or the `build.toml`. This presents several limitations which may be alleviated the this proposal.
+
+Often times `BOM` describes the artifacts present in the layer. Having this present in the `launch.toml` or `build.toml` which are ephemeral between builds means that the `BOM` has to be generated for each build, even if the layer itself has not changed and is cached. 
+
+This leads to surprising behaviors for users at times when they may be generating a `BOM` as part of the layer generation process and may find that their `BOM` is missing if the `BOM` generation process is not called when the layer is assumed to be cached.
+
+Having the `BOM` be a part of the layer content metadata also means that we can pin-point which layer contributed a BOM entry without having to implement extra logic in downstream buildpacks.
+
+- What is the expected outcome?
+
+The lifecycle is able to generate the final `BOM` by combining the `BOM` entries from top level metadata files like `launch.toml` and `build.toml` and also the ones from `<layer-name>.toml` which handling layer caching correctly i.e. if the layer is cached and re-used in the new build - the `BOM` associated with that layer will also be re-used.
+
+# What it is
+[what-it-is]: #what-it-is
+
+The layer content metadata files would be updated to include - 
+
+```
+[[bom]]
+name = "<dependency name>"
+
+[bom.metadata]
```

Comment:
Should we have `bom.build` and `bom.launch`?

Reply:
Ohh I see - https://github.com/buildpacks/rfcs/pull/151/files#diff-d31996aeb587aab806ae658bfb0199878b47d25545090827fa6b81723ca2d033R59 handles build vs. launch dependencies.

Reply:
Yeah, I imagined if the bom describes a layer then the bom type should be derived from layer type. Makes it easier to reason about.